external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaRDD.scala

```diff
@@ -0,0 +1,220 @@
+    " This should not happen, and indicates that messages may have been lost"
+
+  private def errOvershotEnd(itemOffset: Long, part: KafkaRDDPartition): String =
+    s"Got ${itemOffset} > ending offset ${part.untilOffset} " +
+    s"for topic ${part.topic} partition ${part.partition} start ${part.fromOffset}." +
+    " This should not happen, and indicates a message may have been skipped"
+
+  override def compute(thePart: Partition, context: TaskContext): Iterator[R] = {
+    val part = thePart.asInstanceOf[KafkaRDDPartition]
+    assert(part.fromOffset <= part.untilOffset, errBeginAfterEnd(part))
+    if (part.fromOffset == part.untilOffset) {
+      log.warn("Beginning offset ${part.fromOffset} is the same as ending offset " +
+        s"skipping ${part.topic} ${part.partition}")
+      Iterator.empty
+    } else {
+      new KafkaRDDIterator(part, context)
+    }
+  }
+
+  private class KafkaRDDIterator(
+      part: KafkaRDDPartition,
+      context: TaskContext) extends NextIterator[R] {
+
+    context.addTaskCompletionListener{ context => closeIfNeeded() }
+
+    log.info(s"Computing topic ${part.topic}, partition ${part.partition} " +
+      s"offsets ${part.fromOffset} -> ${part.untilOffset}")
+
+    val kc = new KafkaCluster(kafkaParams)
+    val keyDecoder = classTag[U].runtimeClass.getConstructor(classOf[VerifiableProperties])
+      .newInstance(kc.config.props)
+      .asInstanceOf[Decoder[K]]
+    val valueDecoder = classTag[T].runtimeClass.getConstructor(classOf[VerifiableProperties])
+      .newInstance(kc.config.props)
+      .asInstanceOf[Decoder[V]]
+    val consumer = connectLeader
+    var requestOffset = part.fromOffset
+    var iter: Iterator[MessageAndOffset] = null
+
+    // The idea is to use the provided preferred host, except on task retry atttempts,
+    // to minimize number of kafka metadata requests
+    private def connectLeader: SimpleConsumer = {
+      if (context.attemptNumber > 0) {
+        kc.connectLeader(part.topic, part.partition).fold(
+          errs => throw new SparkException(
+            s"Couldn't connect to leader for topic ${part.topic} ${part.partition}: " +
+              errs.mkString("\n")),
+          consumer => consumer
+        )
+      } else {
+        kc.connect(part.host, part.port)
+      }
+    }
+
+    private def handleFetchErr(resp: FetchResponse) {
+      if (resp.hasError) {
+        val err = resp.errorCode(part.topic, part.partition)
+        if (err == ErrorMapping.LeaderNotAvailableCode ||
+          err == ErrorMapping.NotLeaderForPartitionCode) {
+          log.error(s"Lost leader for topic ${part.topic} partition ${part.partition}, " +
+            s" sleeping for ${kc.config.refreshLeaderBackoffMs}ms")
+          Thread.sleep(kc.config.refreshLeaderBackoffMs)
+        }
+        // Let normal rdd retry sort out reconnect attempts
+        throw ErrorMapping.exceptionFor(err)
+      }
+    }
+
```

Comment:
I'm not saying the specification is inherently broken, I'm saying the
specification is this: give a range of offsets, and they will be processed
by spark.

Of course if you give it a range of offsets that no longer exist in Kafka
you're going to get an error.  Silently processing a different range of
offsets than specified by the client would be horrible.  That's what would
break the specification.

I don't see why this is something that needs additional spark specific
documentation.  Its a problem of misusing Kafka, and is documented by
Kafka.  If you're running into that error all the time, either consume
faster or set your log retention longer.
On Jan 31, 2015 10:22 AM, "Dibyendu Bhattacharya" <notifications@github.com>
wrote:

> In
> external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaRDD.scala
> <https://github.com/apache/spark/pull/3798#discussion_r23889938>:
>
> > +    }
> > +
> > +    private def handleFetchErr(resp: FetchResponse) {
> > +      if (resp.hasError) {
> > +        val err = resp.errorCode(part.topic, part.partition)
> > +        if (err == ErrorMapping.LeaderNotAvailableCode ||
> > +          err == ErrorMapping.NotLeaderForPartitionCode) {
> > +          log.error(s"Lost leader for topic ${part.topic} partition ${part.partition}, " +
> > +            s" sleeping for ${kc.config.refreshLeaderBackoffMs}ms")
> > +          Thread.sleep(kc.config.refreshLeaderBackoffMs)
> > +        }
> > +        // Let normal rdd retry sort out reconnect attempts
> > +        throw ErrorMapping.exceptionFor(err)
> > +      }
> > +    }
> > +
>
> If the KafkaRdd specification is broken , then probably you need to
> document it how client code need to handle. In our Production Kafka
> cluster, this error not very rare and not very rare for others as far I
> know who uses Kafka in Prod.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/3798/files#r23889938>.
>